### PR TITLE
Feature/performance tests

### PR DESCRIPTION
**Pull request by markysoft**
Add simple parameterised JMeter test for performance testing

### CODE REVIEW

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles succeeded (http://profiles-pr-192.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles succeeded (http://profiles-pr-192.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles succeeded (http://profiles-pr-192.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles succeeded (http://profiles-pr-192.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles succeeded (http://profiles-pr-192.dev.beta.nhschoices.net)

File: test/performance/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Profiles Performance Test
```
Comment (st3v3nhunt) [inline test/performance/README.md]: Could probably just have the heading as `Performance Testing`

File: test/performance/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+The `profiles.jmx` [JMeter](http://jmeter.apache.org/) test will call the Profiles application for a number of different surgeries.
+Several parameters are exposed to customise the base test.
+
+
+| Parameter                         | Description                                                        | Default               |
+|:---------------------------------|:-------------------------------------------------------------------|-----------------------|
+| `hostname`                       | URL of server to test                                              | profiles.dev.beta.nhschoices.net|
```
Comment (beseven) [inline test/performance/README.md]: Usually, performance tests are done on the staging environment as it's deemed closer to the production configuration. In our case, it should have multiple images per app and theoretically, that should take the load.

File: test/performance/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+| `users`                          | Simulated number of concurrent users                               | 5                  |
+| `rampup`                         | Time in seconds to ramp up to total number of users                | 20                 |
+| `duration`                       | Time in seconds to run test                                        | 120                |
+| `throughput`                     | target throughput in samples per minute                            | 120                |
+| `csvfile`                        | csv data file of choices ID to use, either `ids_100.csv`, `ids_1200.csv`, or `ids_all.csv` (9000+), allowing a number of different GP pages to be visited during the test | ids_100.cs         |
```
Comment (st3v3nhunt) [inline test/performance/README.md]: Missing `v` from default value `ids_100.cs` ➡️  `ids_100.csv`

File: test/performance/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+## Running the Test
+
+The test may be loaded into the JMeter application, or the file may be run from the command line locally by changing into the `test/performance` folder and entering:
+
+/usr/local/bin/jmeter/jmeter -n -t  ./profiles.jmx -Jhostname=*hostname* -Jusers=*users* -Jrampup=*rampup* -Jduration=*duration* -Jthroughput=*throughput* Jcsvfile=*csvfile* -l profiles.jtl
+
+
+Where the parameters in italics can be substituted with the desired value, or removed entirely to use the default.
+
+*Note:* The path to JMeter may differ depending on your installation location.
```
Comment (beseven) [inline test/performance/README.md:25]: Is the approach of running the perf-tests in the container not the same anymore? https://github.com/nhsuk/nhsuk-rancher-templates/compare/feature/pharmacy-finder-perf-tests?expand=1 and if so, how is this approach better?

File: test/performance/README.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+| `rampup`                         | Time in seconds to ramp up to total number of users                | 20                 |
+| `duration`                       | Time in seconds to run test                                        | 120                |
+| `throughput`                     | target throughput in samples per minute                            | 120                |
+| `csvfile`                        | csv data file of choices ID to use, either `ids_100.csv`, `ids_1200.csv`, or `ids_all.csv` (9000+), allowing a number of different GP pages to be visited during the test | ids_100.cs         |
+
+## Running the Test
+
+The test may be loaded into the JMeter application, or the file may be run from the command line locally by changing into the `test/performance` folder and entering:
+
+/usr/local/bin/jmeter/jmeter -n -t  ./profiles.jmx -Jhostname=*hostname* -Jusers=*users* -Jrampup=*rampup* -Jduration=*duration* -Jthroughput=*throughput* Jcsvfile=*csvfile* -l profiles.jtl
```
Comment (st3v3nhunt) [inline test/performance/README.md:20]: Surrounding this with backticks would help differentiate the command from the text.

Perhaps have a script with the default values in `./scripts/test-perf` to allow easy running locally?

**Review by st3v3nhunt — COMMENTED** [review]
A few comments on the `README`. I'll be looking at the test next.

File: test/performance/profiles.jmx [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    </TestPlan>
+    <hashTree>
+      <ThreadGroup guiclass="ThreadGroupGui" testclass="ThreadGroup" testname="Profiles Thread Group" enabled="true">
+        <stringProp name="ThreadGroup.on_sample_error">continue</stringProp>
+        <elementProp name="ThreadGroup.main_controller" elementType="LoopController" guiclass="LoopControlPanel" testclass="LoopController" testname="Loop Controller" enabled="true">
+          <boolProp name="LoopController.continue_forever">false</boolProp>
+          <intProp name="LoopController.loops">-1</intProp>
+        </elementProp>
+        <stringProp name="ThreadGroup.num_threads">${__P(users,5)}</stringProp>
+        <stringProp name="ThreadGroup.ramp_time">${__P(rampup,20)}</stringProp>
```
Comment (st3v3nhunt) [inline test/performance/profiles.jmx:21]: Why a 20 second ramp-up period?

File: test/performance/profiles.jmx [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+      <stringProp name="TestPlan.user_define_classpath"></stringProp>
+    </TestPlan>
+    <hashTree>
+      <ThreadGroup guiclass="ThreadGroupGui" testclass="ThreadGroup" testname="Profiles Thread Group" enabled="true">
+        <stringProp name="ThreadGroup.on_sample_error">continue</stringProp>
+        <elementProp name="ThreadGroup.main_controller" elementType="LoopController" guiclass="LoopControlPanel" testclass="LoopController" testname="Loop Controller" enabled="true">
+          <boolProp name="LoopController.continue_forever">false</boolProp>
+          <intProp name="LoopController.loops">-1</intProp>
+        </elementProp>
+        <stringProp name="ThreadGroup.num_threads">${__P(users,5)}</stringProp>
```
Comment (st3v3nhunt) [inline test/performance/profiles.jmx:20]: Why 5 users?

**Review by st3v3nhunt — CHANGES_REQUESTED** [review]
Why is there no use of a Cookie Manager?

I've made some tweaks in a PR #193

**Comment by st3v3nhunt** [comment]
@markysoft Can there just be a single data file of all of the ids?

File: test/performance/README.md (hunk 2 above)
Comment (markysoft) [reply test/performance/README.md]: Good spot, it runs against staging on the triggered performance tests, I'll update the default.

File: test/performance/README.md (hunk 4 above)
Comment (markysoft) [reply test/performance/README.md:25]: The goal of this PR is to add a JMeter tests to be used by the Teamcity performance dashboards. The information in the readme is to describe the parameters used, and provide assistance if you want to run them locally. 

It's not meant to replace any other approaches, just to describe how you might run them if you'd like to.

File: test/performance/README.md (hunk 5 above)
Comment (markysoft) [reply test/performance/README.md:20]: You can either have italics from the *, or back ticks, they don't seem to play well together . I think italicising the variable names is better than the code formatting.

The scope of this work is to provide a JUnit test for the Teamcity performance checks. Providing a way to run performance tests locally is a big topic, as already mentioned in @beseven 's comment above. 

In my mind the only reason to run them locally is when editing the JMX file.

File: test/performance/profiles.jmx (hunk 6 above)
Comment (markysoft) [reply test/performance/profiles.jmx:21]: These figures are small defaults used when running locally. On the TeamCity build they are overridden with longer times, more users, etc with scenario settings suggested by the performance specialists in line with other projects.

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles succeeded (http://profiles-pr-192.dev.beta.nhschoices.net)

**Comment by markysoft** [comment]
RE: having a single data file - the profiles data changes on a daily basis. If a practice is removed or changes ID then the test will return a 404, which is considered a failure on the performance test, so using the full list can give false negatives.

One approach to mitigate this would be to request a list of valid IDs before the test starts.
For now a small curated list of IDs is provided, with the other data sets available for more intensive testing.

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles succeeded (http://profiles-pr-192.dev.beta.nhschoices.net)

File: test/performance/README.md (hunk 2 above)
Comment (st3v3nhunt) [reply test/performance/README.md]: @markysoft Did you intend to update the default `hostname`?

File: test/performance/README.md (hunk 5 above)
Comment (st3v3nhunt) [reply test/performance/README.md:20]: Damn back ticks 🤛

**Comment by c2s-dev** [comment]
:rocket: deployment of nhsuk/profiles succeeded (http://profiles-pr-192.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of nhsuk/profiles succeeded (http://profiles-pr-192.dev.beta.nhschoices.net)

**Review by st3v3nhunt — APPROVED** [review]

**Pull request merged**